<!DOCTYPE html>
<meta charset="utf-8">
    <style>
        body {
          font-family: 'Open Sans', sans-serif;
        }
        .node rect {
          stroke: black;
          stroke-width: 1.5px;
        }

        .node {
          font-size: 24px;
        }

        .link {
          fill: none;
          stroke: #333;
          stroke-width: 6px;
        }

        .tick text {
          font: 20px;
        }

        .axis line, .axis path {
          fill: none;
          stroke: #000;
        }

        .select2-container {
          width: 400px;
          font: 18px;
        }
        
        .container {
          width: 800px;
          margin-left: auto;
          margin-right: auto;
          padding-top: 30px;
        }

        .chart h1 {
          margin-bottom: 0px;
          margin-left: 10px;
        }

        div.description {
            font-size: 14px;
        }
    </style>
    <link href="//cdnjs.cloudflare.com/ajax/libs/select2/4.0.1/css/select2.min.css" rel="stylesheet" />
    <link href='https://fonts.googleapis.com/css?family=Open+Sans' rel='stylesheet' type='text/css'>
<body>
    <div class="container">
      <div class='description'>
        <h2>Human mesoderm gene expression atlas</h2>
        <p>This is a resource of different genes expressed across multiple successive stages of human mesoderm development, as measured through bulk-population RNA-seq. To access the expression pattern of a given gene, click the drop down box below and select the gene name of interest or type it in manually (<a href="http://www.ncbi.nlm.nih.gov/gene">NCBI official gene symbols</a> are accepted).</p>

        <p>In the schematized mesoderm developmental lineage hierarchy below, each box represents a given cell-type and the intensity of shading represents the expression of a given gene in that cell-type, with respect to the heatmap scale on the right (where log2 gene expression is shown). Expression of each gene is normalized such that within all mesoderm lineages examined, the cell-type with highest expression is assigned the maximum expression value.</p>

          <p>All data are derived from bulk-population RNA-sequencing of H7 human embryonic stem cells and their mesodermal progeny. Detailed experimental methods pertaining to stem-cell differentiation, sequencing library preparation and computational analysis are described in the Supplementary Experimental Procedures accompanying Loh & Chen et al. (2016).
          </p>

          <p>For ease of use, we have also prepared a spreadsheet with TPM values for each gene, augmented with the following information on each gene:
            <ol>
                <li> whether the gene product is present on the cell surface (GO code GO:0009986) </li>
                <li> for each pair of adjacent conditions, whether the gene was differentially expressed between those conditions </li>
                <li> the shrunken log-fold-change for that gene between those conditions. </li>
            </ol>
            (2) and (3) are called by DESeq2 (Love, Huber, Anders, Genome Biology, 2014).
            To make the spreadsheet easier to manipulate, we have only included genes that are differentially expressed in at least one pair of conditions.
            The spreadsheet can be found <a href="https://figshare.com/articles/Annotated_bulk-population_RNA-seq_data_on_a_mesoderm_differentiation_dataset/3842835">here</a>.
            The raw data can be found in SRA (accession number SRP073808), with full TPM spreadsheets in GEO (accession number GSE85066).
          </p>

          <h3>Abbreviations:</h3>
          <p>Human embryonic stem cells (hESCs); Anterior primitive streak (Ant PS); Mid primitive streak (Mid PS); DLL1-sorted paraxial mesoderm (Parax Meso); Lateral mesoderm (Lat Meso); Early somite (Erly Smte); Dermomyotome (Dermomyo); PDGFRA-sorted sclerotome (Sclero); GARP-sorted cardiac mesoderm (Card Meso).
          </p>

          <h3>Contact:</h3>
          <p><a href="mailto:pangwei@cs.stanford.edu">pangwei@cs.stanford.edu</a>,
             <a href="mailto:zhenghao@cs.stanford.edu">zhenghao@cs.stanford.edu</a></p>

          <h3>Citations:</h3>
          <a href="http://www.cell.com/cell/abstract/S0092-8674(16)30740-1">
              <p style='font-size:14px'>Loh & Chen et al. (2016). Mapping the pairwise choices leading from pluripotency to human bone, heart and other mesoderm cell-types. Cell.</p>
          </a>
          <a href="http://www.nature.com/articles/sdata2016109">
              <p style='font-size:14px'> Koh et al. (2016). An atlas of transcriptional, chromatin accessibility, and surface marker changes in human mesoderm development. Scientific Data.</p>
          </a>
          
          <!--TODO:
          (insert bibliographic information/PubMed ID when confirmed). 

          Representative Examples: (Maybe 2-3 heatmaps to show nice lineage-specific genes and to emphasize the usefulness of the atlas)
          -->
      </div>
      <select class="gene-select" style="width:200px"></select>
      <div class="chart">
        <h1 class="title"></h1>
      </div>
    </div>

<script src="//d3js.org/d3.v3.min.js" charset="utf-8"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/underscore.js/1.8.3/underscore-min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.0.0-beta1/jquery.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/select2/4.0.1/js/select2.min.js"></script>
<script>

var width = 800,
    height = 700,
    ymax = 1;

var treeLayout = d3.layout.tree()
    .size([width - 160, height - 100]);

var svg = d3.select(".chart").append("svg")
    .attr("width", width)
    .attr("height", height); 


var tree = 
{'cellType': 'H7 hESC',
 'label': 'hESC',
'children': [

    {'cellType': 'APS',
     'label': 'Ant PS',
     'children': [

        {'cellType': 'DLL1Pos PXM',
         'label': 'Parax Meso',
         'children': [

            {'cellType': 'Somite',
             'label': 'Erly Somte',
             'children': [

                {'cellType': 'Dermomyotome',
                 'label': 'Dermo- myo',
                 'children': []},

                {'cellType': 'Sclerotome',
                 'label': 'Sclero',
                 'children': []}
            ]}
        ]}
    ]},

    {'cellType': 'MPS',
     'label': 'Mid PS',
     'children': [

        {'cellType': 'LatM',
         'label': 'Lat Meso',
         'children': [

            {'cellType': 'H7 Cardiac',
             'label': 'Card Meso',
             'children': []}
        ]}
    ]}
]};

var colorScale = d3.scale.linear()
    .domain([0, 0.5 * ymax, ymax])
    .range(["white", "#FFB03B", "#B12800"]);

var buildLegend = function(colorScale, parentElem, width, height) {
    var ymax = Math.max(...colorScale.domain());
    var legend_data = [];
    for (var i = 0; i < height; i++) {
      legend_data.push(i / height * ymax); 
    }

    var legend_scale = d3.scale.linear()
      .domain([ymax, 0])
      .range([0, height]);

    var legend_axis = d3.svg.axis()
      .scale(legend_scale)
      .orient("right")
      .tickValues(colorScale.domain());

    var key = parentElem.attr("width", width * 2).attr("height", height * 2);

    key.append("g")
      .attr("class", "legend axis")
      .attr("transform", "translate(" + (5 + width / 2) + ", 30)")
      .call(legend_axis)
    .selectAll(".legend_cell")
      .data(legend_data)
      .enter().append("rect")
      .attr("y", function(d, i) { return i;})
      .attr("x", -width / 2)
      .attr("width", width / 2)
      .attr("height", 1)
      .attr("fill", function(d) { return colorScale(ymax - d); });
};

var elbow = function(d, i) {
  var alpha = 0.5;
  return "M" + d.source.x + "," + d.source.y
      + "V" + (d.source.y * (1 - alpha) + d.target.y * alpha)
      + "H" + d.target.x
      + "V" + d.target.y;
}

var renderTree = function(tree) {
  var nodes = treeLayout.nodes(tree);
  var tree = svg.append("g")
    .attr("transform", "translate(-50, 50)");

  var link = tree.selectAll("path.link")
      .data(treeLayout.links(nodes))
    .enter().append("path")
      .attr("class", "link")
      .attr("d", elbow);

  var node = tree.selectAll("g.node")
      .data(nodes)
    .enter().append("g")
      .attr("class", "node")
      .attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; })

  node.append("rect")
      .attr("width", "90")
      .attr("height", "90")
      .attr("fill", "white")
      .attr("transform", "translate(-45, -45)")

  node.append("text")
      .attr("dx", 0)
      .attr("dy", 0.2)
      .attr("text-anchor", "middle")
      .text(function(d) { return d.label; })
      .attr("fill", "black");

  node.selectAll("text")
    .call(wrap, 90);

  var w = 80, h = 200;
  var legend = svg.append("g").attr("width", w * 2).attr("height", h * 2)
    .attr("transform", "translate(" + (width - 160) + ", 50)");
  buildLegend(colorScale, legend, w, h);
}

d3.csv("assets/allData.csv", function (x) {
  var augmentTree = function(node) {
    node.geneValues = _.findWhere(x, {'cellType': node.cellType});
    for (var i = 0, len = node.children.length; i < len; i++) {
      augmentTree(node.children[i]);
    }
  };
  augmentTree(tree);
  renderTree(tree);
  
  var genes = [];
  _.each(_.keys(x[0]), function (gene) {
    if (gene != "cellType")
      genes.push({"id": gene, "text": gene});
  });
    
  $(".gene-select").select2({data: genes})
    .on("select2:select", function (e) {
      var gene = this.value;
      svg.selectAll(".node rect")
        .attr("fill", function(d) {return colorScale(d.geneValues[gene])})
      svg.selectAll(".node text")
        .attr("fill", function(d) {return (d.geneValues[gene] > 0.8) ? "white" : "black"})
      d3.select(".title")
        .text(gene);
    });
});

function wrap(text, width) {
  text.each(function() {
    var text = d3.select(this),
        words = text.text().split(/\s+/).reverse(),
        word,
        line = [],
        lineNumber = 0,
        lineHeight = 1.1, // ems
        y = text.attr("y"),
        dy = parseFloat(text.attr("dy")),
        tspan = text.text(null)
                  .append("tspan")
                  .attr("x", 0)
                  .attr("y", y)
                  .attr("dy", dy + "em")
                  .attr("data-line-number", 0);
    
    while (word = words.pop()) {
      line.push(word);
      tspan.text(line.join(" "));
      if (tspan.node().getComputedTextLength() > width) {
        line.pop();
        tspan.text(line.join(" "));
        line = [word];
        tspan = text.append("tspan")
            .attr("x", 0)
            .attr("y", y)
            .attr("dy", ++lineNumber * lineHeight + dy + "em")
            .text(word)
            .attr("data-line-number", lineNumber);
      }
    }
    
    text.select("tspan")
      .attr("dy", (-(lineNumber / 2) * lineHeight + dy) + "em");
    
  });
}

</script>
</body>
</html>
